File: model/src/main/java/io/github/chriszhong/model/organization/event/ModeratesEvent.java
package io.github.chriszhong.model.organization.event;

import io.github.chriszhong.message.M;
import io.github.runtimemodels.chazm.entity.Attribute;
import io.github.runtimemodels.chazm.entity.Pmf;
import io.github.runtimemodels.chazm.id.UniqueId;
import io.github.runtimemodels.chazm.relation.Moderates;

import java.util.Objects;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain ModeratesEvent} class indicates that there is an update about a {@linkplain Moderates} relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class ModeratesEvent extends AbstractEvent {

	private static final long serialVersionUID = 273935856408749575L;
	private final UniqueId<Pmf> pmfId;
	private final UniqueId<Attribute> attributeId;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	ModeratesEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Moderates moderates) {
		super(category);
		pmfId = moderates.getPmf().getId();
		attributeId = moderates.getAttribute().getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Pmf}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Pmf> getPmfId() {
		return pmfId;
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents an {@linkplain Attribute}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Attribute> getAttributeId() {
		return attributeId;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof ModeratesEvent) {
			final ModeratesEvent event = (ModeratesEvent) object;
			return super.equals(event) && getPmfId().equals(event.getPmfId()) && getAttributeId().equals(event.getAttributeId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			hashCode = Objects.hash(getCategory(), getPmfId(), getAttributeId());
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = M.EVENT_WITH_2_IDS.get(super.toString(), getPmfId(), getAttributeId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/ModeratesEvent.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.AttributeId
import runtimemodels.chazm.api.entity.Pmf
import runtimemodels.chazm.api.entity.PmfId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Moderates
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [ModeratesEvent] class indicates that there is an update about a [Moderates] relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class ModeratesEvent @Inject internal constructor(
    category: EventType,
    moderates: Moderates
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Pmf].
     *
     * @return a [UniqueId].
     */
    val pmfId: PmfId = moderates.pmf.id
    /**
     * Returns a [UniqueId] that represents an [Attribute].
     *
     * @return a [UniqueId].
     */
    val attributeId: AttributeId = moderates.attribute.id

    override fun equals(other: Any?): Boolean {
        if (other is ModeratesEvent) {
            return super.equals(other) && pmfId == other.pmfId && attributeId == other.attributeId
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, pmfId, attributeId)

    override fun toString(): String = M.EVENT_WITH_2_IDS[super.toString(), pmfId, attributeId]

    companion object {
        private const val serialVersionUID = 273935856408749575L
    }

}
